security/pkg/k8s/configmap/configmap.go

```diff
@@ -17,35 +17,131 @@ package configmap
 import (
 	"context"
 	"fmt"
+	"sort"
+	"strings"
+	"sync"
 	"time"
 
-	"istio.io/pkg/log"
-
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
 	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/pkg/log"
+
+	"istio.io/istio/security/pkg/pki/util"
 )
 
 var configMapLog = log.RegisterScope("configMapController", "ConfigMap controller log", 0)
 
 const (
 	IstioSecurityConfigMapName = "istio-security"
 	CATLSRootCertName          = "caTLSRootCert"
+
+	// ConfigMap label key that indicates that the ConfigMap contains additional trusted CA roots.
+	ExtraTrustAnchorsLabel = "security.istio.io/extra-trust-anchors"
+
+	// periodically refresh the primaryTrustAnchor from the CA.
+	PrimaryTrustAnchorRefreshTimeout = time.Second
 )
 
+// CertificateAuthority contains methods to be supported by a CA.
+type CertificateAuthority interface {
+	// Sign generates a certificate for a workload or CA, from the given CSR and TTL.
+	// TODO(myidpt): simplify this interface and pass a struct with cert field values instead.
+	Sign(csrPEM []byte, subjectIDs []string, ttl time.Duration, forCA bool) ([]byte, error)
+	// SignWithCertChain is similar to Sign but returns the leaf cert and the entire cert chain.
+	SignWithCertChain(csrPEM []byte, subjectIDs []string, ttl time.Duration, forCA bool) ([]byte, error)
+	// GetCAKeyCertBundle returns the KeyCertBundle used by CA.
+	GetCAKeyCertBundle() util.KeyCertBundle
+}
+
 // Controller manages the CA TLS root cert in ConfigMap.
 type Controller struct {
```

Comment:
I think we should put the functions of retrieving extra trust anchors into a separate file, maybe name it as "extraTrustAnchors". The original one can be renamed as "citadelRoot".